Presence of bearer token should cancel exec action

anderseknert opened the pull request:
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
If a bearer token is present in a request/config, the exec credential plugin should accept that as the chosen method of authentication. Judging by an [earlier comment in exec.go](https://github.com/kubernetes/kubernetes/blob/c18bc7e9f7a27d7d197053d98c52896e1dc3bb3e/staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go#L217), this was already the intention. This would however not work since `UpdateTransportConfig` would set the `GetCert` callback which would then get called by the transport, triggering the exec plugin action even with a token present in the request. See linked issue for further details.

**Which issue(s) this PR fixes**:
Fixes #87369

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fixed ambiguous behavior when bearer token (kubectl --token=..) and an exec credential plugin was configured in the same context - the bearer token now takes precedence.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

Signed-off-by: Anders Eknert <anders.eknert@bisnode.com>

anderseknert committed:
Presence of bearer token should cancel exec action

If a bearer token is present in a request, the exec credential plugin should accept that as the chosen method of authentication. Judging by an [earlier comment in exec.go](https://github.com/kubernetes/kubernetes/blob/c18bc7e9f7a27d7d197053d98c52896e1dc3bb3e/staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go#L217), this was already intended. This would however not work since UpdateTransportConfig would set the GetCert callback which would then get called by the transport, triggering the exec plugin action even with a token present in the request. See linked issue for further details.

See #87369 for further details.

Signed-off-by: Anders Eknert <anders.eknert@bisnode.com>

anderseknert commented:
/assign @liggitt

liggitt commented:
/unassign
/assign @enj

fedebongio commented:
/remove-sig api-machinery

enj commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:
```diff
@@ -182,6 +182,10 @@ func (a *Authenticator) UpdateTransportConfig(c *transport.Config) error {
 		return &roundTripper{a, rt}
 	})
 
+	if c.HasTokenAuth() {
+		return nil
+	}
+
```
Please add a comment explaining in detail (i.e. https://github.com/kubernetes/kubernetes/issues/87369#issuecomment-585579312) why we basically need the same check twice.

anderseknert replied:
Done.

anderseknert replied:
Done.

enj reviewed (commented):
LGTM, but I would like to see a `kubectl get --token $TOKEN pods` style test to prove that this actually overrides the `exec` plugin in the way that we expect it to.

enj commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:
```diff
@@ -182,6 +182,10 @@ func (a *Authenticator) UpdateTransportConfig(c *transport.Config) error {
 		return &roundTripper{a, rt}
 	})
 
+	if c.HasTokenAuth() {
+		return nil
+	}
```
Thinking on this a bit more, maybe this should be the first line in this function.  As it stands, it attempts to configure token auth via the `c.Wrap` call but then skips the cert auth.

It may make more sense for `UpdateTransportConfig` to do nothing when `c.HasTokenAuth()` returns `true` instead of doing "half" of what it is supposed to do.  We still want the `req.Header.Get("Authorization") != ""` logic below because round trippers can be arbitrarily stacked via `c.Wrap`.

anderseknert replied:
Agreed - moved to top of function.

anderseknert replied:
Agreed - done.

enj commented:
/ok-to-test
/priority important-longterm
/milestone v1.19

anderseknert commented:
> LGTM, but I would like to see a `kubectl get --token $TOKEN pods` style test to prove that this actually overrides the `exec` plugin in the way that we expect it to.

I had never written that type of test in this code base before so it took some time to get something working. First iteration of such test did _not_ trigger the exec credential plugin action so I struggled for quite some time trying to figure out what would cause the behavior to be different with the running test server vs. my own servers (including kind) where it was easy to trigger. Turns out the GetCert callback is only called if the server advertised client certificate authentication capabilities in the TLS handshake! So with that added to the test server configuration the exec action is indeed triggered even with `--token=xxx` provided in the request. 

You can verify by running the integration test (`make test-cmd WHAT="exec_credentials"`) with and without the early return in the `UpdateTransportConfig` function - the kubectl call is done with `--token=admin-token` and should thus not trigger the exec action.

Since I hadn't written this type of test before I probably did a couple of rookie mistakes - I tried looking at other tests of the same type for inspiration, but any feedback is appreciated :)

anderseknert commented:
@enj let me know if there's anything I need to add to this. Not sure why the bot wants me to (re-)assign this, am I mean to do that?

enj commented:
> @enj let me know if there's anything I need to add to this. Not sure why the bot wants me to (re-)assign this, am I meant to do that?

I will review shortly.  The bot is saying "you do not have an approver of this code assigned to this PR and you will need one of those to merge it" - do not worry about that, I will find the right person to get this merged.

anderseknert commented:
A perhaps easier way to demonstrate/reproduce can be found here: https://github.com/anderseknert/87369-reproduce

If that would be better to include as a test case I'd be happy to replace the current one, but I might need some pointers where to put it.

enj commented on test/cmd/legacy-script.sh:
```diff
@@ -29,6 +29,7 @@ KUBE_ROOT=$(dirname "${BASH_SOURCE[0]}")/../..
 # source "${KUBE_ROOT}/hack/lib/test.sh"
 source "${KUBE_ROOT}/test/cmd/apply.sh"
 source "${KUBE_ROOT}/test/cmd/apps.sh"
+source "${KUBE_ROOT}/test/cmd/auth.sh"
```
```suggestion
source "${KUBE_ROOT}/test/cmd/authentication.sh"
```

enj replied:
You need to call `run_exec_credentials_tests` in this file to actually run the tests (see other examples later in this file).

The [CI job](https://prow.k8s.io/pr-history/?org=kubernetes&repo=kubernetes&pr=91745) `pull-kubernetes-integration` should show the test running in the [raw build log](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/91745/pull-kubernetes-integration/1273332842755002372).

anderseknert replied:
Oh, right, good catch!

enj commented on hack/make-rules/test-cmd.sh:
```diff
@@ -69,6 +69,9 @@ function run_kube_apiserver() {
     --storage-media-type="${KUBE_TEST_API_STORAGE_TYPE-}" \
     --cert-dir="${TMPDIR:-/tmp/}" \
     --service-cluster-ip-range="10.0.0.0/24" \
+    --client-ca-file=hack/testdata/ca.crt \
+    --tls-cert-file=hack/testdata/server.crt \
+    --tls-private-key-file=hack/testdata/server.key \
```
@liggitt any concerns with adding this?

enj reviewed (changes requested):
> A perhaps easier way to demonstrate/reproduce can be found here: https://github.com/anderseknert/87369-reproduce

The approach you have in this PR seems fine to me.  Go ahead and squash after addressing comments.

enj commented on test/cmd/auth.sh:
```diff
@@ -0,0 +1,69 @@
+  # Pre-condition: None
+  # Command
+  kubectl "${kube_flags_with_token[@]:?}" --kubeconfig="${TMPDIR:-/tmp}"/invalid_exec_plugin.yaml get namespaces 2>"${EXEC_ERROR_FILE}" || true
+  if grep -q "decoding stdout: couldn't get version/kind; json parse error" "${EXEC_ERROR_FILE}"; then
```
I think it would be safer long term to check for the positive case, i.e. run `kubectl get namespace kube-system -o name --token=...` and make sure the command runs successfully and the output is `namespace/kube-system`.

anderseknert replied:
Changed along with other suggested improvements. Commits squashed.

enj commented:
I "undid" the fix from this PR and manually ran the test to confirm that it fails:

```
Recording: run_exec_credentials_tests
Running command: run_exec_credentials_tests

+++ Running case: test-cmd.run_exec_credentials_tests 
+++ working dir: .../kubernetes/kubernetes/src/github.com/kubernetes/kubernetes
+++ command: run_exec_credentials_tests
+++ [0622 22:06:40] Testing kubectl with configured exec credentials plugin
I0622 22:06:40.109456    5031 log.go:181] http: TLS handshake error from 127.0.0.1:57212: EOF
+++ [0622 22:06:40] exec credential plugin triggered despite kubectl being called with --token
error: couldn't read version from server: Get "https://127.0.0.1:6443/version?timeout=32s": decoding stdout: couldn't get version/kind; json parse error: json: cannot unmarshal string into Go value of type struct { APIVersion string "json:\"apiVersion,omitempty\""; Kind string "json:\"kind,omitempty\"" }
+++ exit code: 1
+++ error: 1
Error when running run_exec_credentials_tests
```

enj commented:
CI failures look unrelated / strange:

/retest

/lgtm

/assign @liggitt
For approval

enj commented:
/lgtm cancel

Integration failure is related:

```
run_kubectl_config_set_cluster_tests expand_less | 0s
-- | --
base64: /tmp/apiserver.crt: No such file or directory !!! [0625 18:35:49] Call tree: !!! [0625 18:35:49]  1: /home/prow/go/src/k8s.io/kubernetes/test/cmd/../../third_party/forked/shell2junit/sh2ju.sh:47 run_kubectl_config_set_cluster_tests(...) !!! [0625 18:35:49]  2: /home/prow/go/src/k8s.io/kubernetes/test/cmd/../../third_party/forked/shell2junit/sh2ju.sh:112 eVal(...) !!! [0625 18:35:49]  3: /home/prow/go/src/k8s.io/kubernetes/test/cmd/legacy-script.sh:131 juLog(...) !!! [0625 18:35:49]  4: /home/prow/go/src/k8s.io/kubernetes/test/cmd/legacy-script.sh:447 record_command(...) !!! [0625 18:35:49]  5: hack/make-rules/test-cmd.sh:153 runTests(...)
```

Need to check verify as well.

anderseknert commented:
Hi @enj ! Just got a daughter this Monday so I've not had much time to look into this :) 

> Integration failure is related:

Looks like it might be, but I can't see how from here.

> base64: /tmp/apiserver.crt: No such file or directory

Any clue about what might cause this? My test does not delete anything except for its own config file, but as you pointed out before the first iteration of this PR did not even _run_ my test and still these tests were failing. The fix itself seems unlikely to cause this so what remains then would be the new certs put in `testdata` and provided to the integration test API server.. but how that would cause the `apiserver.crt` to vanish I don't know.

enj commented:
> Hi @enj ! Just got a daughter this Monday so I've not had much time to look into this :)

Congrats! 🎉 👶 

If you need to hand this PR off, please let me know.

> > Integration failure is related:
> 
> Looks like it might be, but I can't see how from here.

It is probably not related.  I thought the filenames matched the one you added.

> > base64: /tmp/apiserver.crt: No such file or directory
> 
> Any clue about what might cause this? My test does not delete anything except for its own config file, but as you pointed out before the first iteration of this PR did not even _run_ my test and still these tests were failing. The fix itself seems unlikely to cause this so what remains then would be the new certs put in `testdata` and provided to the integration test API server.. but how that would cause the `apiserver.crt` to vanish I don't know.

I believe the failing test is new in #91077.  Can you rebase your PR on master?  CI might just be doing something weird here.

anderseknert commented:
> > Hi @enj ! Just got a daughter this Monday so I've not had much time to look into this :)
> 
> Congrats! 🎉 👶
> 
> If you need to hand this PR off, please let me know.

Thanks @enj ! 😃 I had a few minutes to spare this evening, and I think I figured it out. Checking the kube-apiserver flag docs and I noticed this:

> --tls-cert-file string
> [...] If HTTPS serving is enabled, and --tls-cert-file and --tls-private-key-file are not provided, a self-signed certificate and key are generated for the public address and saved to the directory specified by --cert-dir.

So, by providing a `--tls-cert-file` and `--tls-private-key-file` I unknowingly disabled the `--cert-dir` which was of course where that other test was looking for the `apiserver.crt`. I would probably have preferred the API server to complain on the presence of both rather than ignoring the `--cert-dir`, but I guess that's an issue for another time :) My test did not rely on any specific `--tls-cert-file` or `--tls-private-key-file` so if these are auto-generated in `cert-dir` just providing the `--client-ca-file` should be enough. Still waiting for the tests to complete as I write this (hence the "should") but running them locally seems to confirm that was the issue. Will check back tomorrow.

anderseknert commented:
OK, so I think that did it - the two failing test suites now remaining look completely unrelated to me.

enj commented on test/cmd/authentication.sh:
```diff
@@ -0,0 +1,66 @@
+  ### Provided --token should take precedence, thus not triggering the (invalid) exec credential plugin
+  # Pre-condition: None
+  # Command
+  output=$(kubectl "${kube_flags_with_token[@]:?}" --kubeconfig="${TMPDIR:-/tmp}"/invalid_exec_plugin.yaml get namespace kube-system -o name || true)
```
Ah I just realized we need to run this command the other way as well:

```bash
output2=$(kubectl --kubeconfig="${TMPDIR:-/tmp}"/invalid_exec_plugin.yaml get namespace kube-system -o name || true)
```

Without the token it should fail with the parse error (which confirms that the `exec` plugin is at least attempting to run).

anderseknert replied:
Added! Needed to use the kube_flags still but without the token or else it would default to the insecure port.

enj commented on hack/make-rules/test-cmd.sh:
```diff
@@ -69,6 +69,7 @@ function run_kube_apiserver() {
     --storage-media-type="${KUBE_TEST_API_STORAGE_TYPE-}" \
     --cert-dir="${TMPDIR:-/tmp/}" \
     --service-cluster-ip-range="10.0.0.0/24" \
+    --client-ca-file=hack/testdata/ca.crt \
```
Is there something we could run in the `run_exec_credentials_tests` check to confirm that the client cert auth is enabled on the API server?  I am not sure if this is observable from the outside.  🤔

enj replied:
Actually I think you can do something like:

`openssl s_client -connect "127.0.0.1:${API_PORT}" -state`

and check to make sure the output contains:

```
Acceptable client certificate CA names
/CN=127.0.0.1
```

enj replied:
Similar approach:

```
$ echo | openssl s_client -connect "127.0.0.1:${API_PORT}" -msg 2>/dev/null | grep CertificateRequest
```

Should output something like:

```
<<< TLS 1.2 Handshake [length 009b], CertificateRequest
```

anderseknert replied:
Indeed, the `openssl s_client` trick is what I used to figure out that the client certificate capabilities of the API server was what triggered the exec credential plugin before. Added as a pre-condition.

enj reviewed (approved):
Minor comments.

anderseknert commented:
Added pre-condition check for client certificate authentication capabilities and a test without `--token` to make sure the plugin is triggered. @enj  - let me know if the last commit looks OK and I'll squash it.

anderseknert commented:
Integration tests not giving the same output on the CI server as on my local machine. Will look into it further tomorrow.

enj commented on hack/lib/util.sh:
```diff
@@ -452,6 +452,22 @@ function kube::util::test_openssl_installed {
+
+  if [[ "${output}" != *"/CN=127.0.0.1"* ]] || [[ "${output}" != *"CN = 127.0.0.1"* ]]; then
+    echo "API server not configured for client certificate authentication"
+    echo "Output of from acceptable client certificate check: ${output}"
```
On CI this looks like:

```
Acceptable client certificate CA names
CN = 127.0.0.1
Requested Signature Algorithms: RSA-PSS+SHA256:ECDSA+SHA256:Ed25519:RSA-PSS+SHA384:RSA-PSS+SHA512:RSA+SHA256:RSA+SHA384:RSA+SHA512:ECDSA+SHA384:ECDSA+SHA512:RSA+SHA1:ECDSA+SHA1
Shared Requested Signature Algorithms: RSA-PSS+SHA256:ECDSA+SHA256:Ed25519:RSA-PSS+SHA384:RSA-PSS+SHA512:RSA+SHA256:RSA+SHA384:RSA+SHA512:ECDSA+SHA384:ECDSA+SHA512
```

anderseknert commented:
OK @enj - now working both locally and on CI server. I've rebased from master and squashed the commits. Quite a process for adding a nil return to a function 😄 But feels good to have this well covered! Thanks for guiding me along the way 👍

enj commented:
> now working both locally and on CI server. I've rebased from master and squashed the commits.

/lgtm

🎉 

> Quite a process for adding a nil return to a function 😄

I think you just described making literally any change to `k/k` xD

> Thanks for guiding me along the way 👍

My pleasure!

enj commented:
/retest

enj commented:
@liggitt this is ready for you now.

liggitt commented on test/cmd/authentication.sh:
```diff
@@ -0,0 +1,83 @@
+  ### Without provided --token, the exec credential plugin should be triggered
+  # Pre-condition: Client certificate authentication enabled on the API server - already checked by positive test above
+  
+  kube_flags_without_token=('-s' "https://127.0.0.1:${SECURE_API_PORT}" '--insecure-skip-tls-verify=true')
```
nit: defining this far away from `kube_flags_with_token` is not ideal... it places assumptions about the secured endpoint/port far away from where we actually start the server

options:
1. define this in the same place we define `kube_flags_with_token` (and maybe rename it to make it clear the distinction is that it is pointing at a secured endpoint, e.g. `kube_flags_secure_without_token`)
2. use `kube_flags_with_token` and then append another `--token=` arg to clear out the token

I have a slight preference for 1

anderseknert replied:
I agree on moving it next to `kube_flags_with_token`, and I'll do so. However, `kube_flags_with_token` also uses the secure port, so I'm not sure about including `secure` in the `without_token` version. Did you think of `kube_flags` which is defined next to it, and uses the insecure port?

anderseknert replied:
Or you meant renaming both of them? Anyway, I've moved them so that they're next to each other, with the `with_token` version simply concatenating the token to the `without_token` values.

liggitt commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:
```diff
@@ -178,6 +178,15 @@ type credentials struct {
+	// also configured to allow client certificates for authentication. For requests
+	// like "kubectl get --token (token) pods" we should assume the intention is to
+	// use the provided token for authentication.
+	if c.HasTokenAuth() {
```
if we already have a client cert/key, should that take precedence as well and prevent calling the exec plugin?

anderseknert replied:
I think that makes sense, yes. This PR addressed a real problem we experienced with our exec plugin using tokens, and the client certificate scenario wasn't something I considered - I don't even know what the current behavior there is to be honest, but I could certainly look into supporting that case as well if you want. With that said, I think the single argument of `--token` makes it a _lot_ more likely to be used ad-hoc than client certificates, which requires something like three(?) file locations to be provided.

liggitt replied:
I think the current state of the PR is coherent w.r.t. token auth. A follow up issue to discuss/resolve treatment of client cert auth is fine with me

anderseknert committed:
Move kube_flags_without_token creation

liggitt reviewed (commented):
/lgtm
/approve
/retest

liggitt commented:
/retest

liggitt commented:
/retest

enj commented:
/retest

liggitt commented:
/retest
